test/core/surface/channel_init_test.cc

```diff
@@ -0,0 +1,172 @@
+            std::vector<std::string>({"foo"}));
+  EXPECT_EQ(GetFilterNames(init, GRPC_SERVER_CHANNEL, ChannelArgs()),
+            std::vector<std::string>());
+}
+
+TEST(ChannelInitTest, DefaultLexicalOrdering) {
+  // ChannelInit defaults to lexical ordering in the absense of other
+  // constraints, to ensure that a stable ordering is produced between builds.
+  ChannelInit::Builder b;
+  b.RegisterFilter(GRPC_CLIENT_CHANNEL, FilterNamed("foo"));
+  b.RegisterFilter(GRPC_CLIENT_CHANNEL, FilterNamed("bar"));
+  b.RegisterFilter(GRPC_CLIENT_CHANNEL, FilterNamed("baz"));
+  auto init = b.Build();
+  EXPECT_EQ(GetFilterNames(init, GRPC_CLIENT_CHANNEL, ChannelArgs()),
+            std::vector<std::string>({"bar", "baz", "foo"}));
+}
+
+TEST(ChannelInitTest, AfterConstraintsApply) {
+  ChannelInit::Builder b;
+  b.RegisterFilter(GRPC_CLIENT_CHANNEL, FilterNamed("foo"));
+  b.RegisterFilter(GRPC_CLIENT_CHANNEL, FilterNamed("bar"))
+      .After({FilterNamed("foo")});
+  b.RegisterFilter(GRPC_CLIENT_CHANNEL, FilterNamed("baz"));
+  auto init = b.Build();
+  EXPECT_EQ(GetFilterNames(init, GRPC_CLIENT_CHANNEL, ChannelArgs()),
+            std::vector<std::string>({"baz", "foo", "bar"}));
+}
+
+TEST(ChannelInitTest, BeforeConstraintsApply) {
+  ChannelInit::Builder b;
+  b.RegisterFilter(GRPC_CLIENT_CHANNEL, FilterNamed("foo"))
+      .Before({FilterNamed("bar")});
+  b.RegisterFilter(GRPC_CLIENT_CHANNEL, FilterNamed("bar"));
+  b.RegisterFilter(GRPC_CLIENT_CHANNEL, FilterNamed("baz"));
+  auto init = b.Build();
+  EXPECT_EQ(GetFilterNames(init, GRPC_CLIENT_CHANNEL, ChannelArgs()),
+            std::vector<std::string>({"baz", "foo", "bar"}));
+}
+
+TEST(ChannelInitTest, PredicatesCanFilter) {
+  ChannelInit::Builder b;
+  b.RegisterFilter(GRPC_CLIENT_CHANNEL, FilterNamed("foo"))
+      .IfChannelArg("foo", true);
+  b.RegisterFilter(GRPC_CLIENT_CHANNEL, FilterNamed("bar"))
+      .IfChannelArg("bar", false);
+  auto init = b.Build();
+  EXPECT_EQ(GetFilterNames(init, GRPC_CLIENT_CHANNEL, ChannelArgs()),
+            std::vector<std::string>({"foo"}));
+  EXPECT_EQ(GetFilterNames(init, GRPC_CLIENT_CHANNEL,
+                           ChannelArgs().Set("foo", false)),
+            std::vector<std::string>({}));
+  EXPECT_EQ(
+      GetFilterNames(init, GRPC_CLIENT_CHANNEL, ChannelArgs().Set("bar", true)),
+      std::vector<std::string>({"bar", "foo"}));
+  EXPECT_EQ(GetFilterNames(init, GRPC_CLIENT_CHANNEL,
+                           ChannelArgs().Set("bar", true).Set("foo", false)),
+            std::vector<std::string>({"bar"}));
+}
+
+TEST(ChannelInitTest, CanAddTerminalFilter) {
+  ChannelInit::Builder b;
+  b.RegisterFilter(GRPC_CLIENT_CHANNEL, FilterNamed("foo"));
+  b.RegisterFilter(GRPC_CLIENT_CHANNEL, FilterNamed("bar")).Terminal();
+  auto init = b.Build();
+  EXPECT_EQ(GetFilterNames(init, GRPC_CLIENT_CHANNEL, ChannelArgs()),
+            std::vector<std::string>({"foo", "bar"}));
+}
+
+TEST(ChannelInitTest, CanAddMultipleTerminalFilters) {
```

Comment:
Please also test the case where there are multiple terminal filters actually selected, which should fail.

Reply:
Doesn't look like you addressed this comment.  Quoting it here in case the github UI makes it hard to fine:

> Please also test the case where there are multiple terminal filters actually selected, which should fail.

Reply:
Looks like we still need a test for the case where there are multiple terminal filters that are actually selected, which should fail.